Classes/common/Encryption/Keychain/CDTEncryptionKeychainProvider.m

```diff
@@ -0,0 +1,78 @@
+//
+//  CDTEncryptionKeychainProvider.m
+//  CloudantSync
+//
+//  Created by Enrique de la Torre Fernandez on 21/04/2015.
+//  Copyright (c) 2015 IBM Cloudant. All rights reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file
+//  except in compliance with the License. You may obtain a copy of the License at
+//  http://www.apache.org/licenses/LICENSE-2.0
+//  Unless required by applicable law or agreed to in writing, software distributed under the
+//  License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+//  either express or implied. See the License for the specific language governing permissions
+//  and limitations under the License.
+//
+
+#import "CDTEncryptionKeychainProvider.h"
+
+#import "TDMisc.h"
+#import "CDTLogging.h"
+
+@interface CDTEncryptionKeychainProvider ()
+
+@property (strong, nonatomic) NSString *password;
+@property (strong, nonatomic) CDTEncryptionKeychainManager *manager;
+
+@end
+
+@implementation CDTEncryptionKeychainProvider
+
+#pragma mark - Init method
+- (instancetype)init { return [self initWithPassword:nil forManager:nil]; }
+
+- (instancetype)initWithPassword:(NSString *)password
+                      forManager:(CDTEncryptionKeychainManager *)manager
+{
+    self = [super init];
+    if (self) {
+        if (password && manager) {
```

Comment:
But I think this case is not exactly the same. In the case of `CDTDatastore`, if a key provider is not supplied, we can not return a datastore and without a datastore, you can do pretty much nothing at all. That is why I favoured an `Assert`.
In the case of a `CDTEncryptionKeychainProvider`, the password as well as the manager are mandatory, so I have to validate them but there are many reason why a password might not be supplied, for example if the user chooses to delegate to the `init` method to check if a password was typed in the keyboard and this is not a reason to raise an exception.
Of course, this is only my point of view, it is OK for me to change the validation but I'd like some agreement about it.

@mikerhodes: What do you think?